src/DefinitionResolver.php

```diff
@@ -479,8 +483,10 @@ public function resolveExpressionNodeToType(Node\Expr $expr): Type
                 if ($expr instanceof Node\Expr\MethodCall) {
                     $fqn .= '()';
                 }
-                $def = $this->index->getDefinition($fqn);
-                if ($def !== null) {
+                if ($def = $this->index->getDefinition($fqn)) {
+                    if ($def->type instanceof Types\This || $def->type instanceof Types\Self_) {
+                        return $this->resolveExpressionNodeToType($expr->var);
+                    }
```

Comment:
Well, unfortunately I could not find a standard for this, and it seems `@return self` is very wide-spread. Maybe because it is the only option defined in the [phpDocumentor site](https://www.phpdoc.org/docs/latest/references/phpdoc/types.html)

As for `this` - I thought I was answering both questions - if you replace `@return self` with `@return $this` in the above example `$def->type` will be an instance of `phpDocumentor\Reflection\Types\This`. I believe this is the correct behavior and will be very important for fluent interfaces - all definitions in the index still use `Types\This`, it is evaluated as needed - hence the check. Maybe I am missing something?